Update documentation for constraints installation (#31882)
mistercrunch/Airflow: docs/apache-airflow/installation/installing-from-pypi.rst

  the problem in `this PR <https://github.com/bazelbuild/rules_python/pull/1166>`_ so it might be that
  newer versions of ``bazel`` will handle it.

Typical command to install airflow from scratch in a reproducible way from PyPI looks like below:

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"

Typically, you can add other dependencies and providers as separate command after the reproducible
installation - this way you can upgrade or downgrade the dependencies as you see fit, without limiting them to
constraints. Good practice for those is to extend such ``pip install`` command with the ``apache-airflow``
pinned to the version you have already installed to make sure it is not accidentally
upgraded or downgraded by ``pip``.

.. code-block:: bash

>     pip install "apache-airflow==|version|" apache-airflow-providers-google=10.1.0

Reviewer: ```suggestion
    pip install "apache-airflow==|version|" apache-airflow-providers-google==10.1.0
```

Those are just examples, see further for more explanation why those are the best practices.

.. note::

   Generally speaking, Python community established practice is to perform application installation in a
   virtualenv created with ``virtualenv`` or ``venv`` tools. You can also use ``pipx`` to install Airflow in a
   application dedicated virtual environment created for you. There are also other tools that can be used

>    to manage your virtualenv installation and you are free to choose how you are managing the environments,

Reviewer: ```suggestion
   to manage your virtualenv installation and you are free to choose how you are managing the environments.
```

Constraints files
'''''''''''''''''

> Why do we need constraints

Reviewer 2: The below suggestion is more of a statement so it doesn't need a `?` at the end.
```suggestion
Why we need constraints
```

Author: yep

==========================

Airflow installation can be tricky because Airflow is both a library and an application.

Libraries usually keep their dependencies open and applications usually pin them, but we should do neither
and both at the same time. We decided to keep our dependencies as open as possible
(in ``setup.cfg`` and ``setup.py``) so users can install different
version of libraries if needed. This means that from time to time plain ``pip install apache-airflow`` will
not work or will produce an unusable Airflow installation.

Reproducible Airflow installation
=================================

In order to have a reproducible installation, we also keep a set of constraint files in the
``constraints-main``, ``constraints-2-0``, ``constraints-2-1`` etc. orphan branches and then we create a tag
for each released version e.g. :subst-code:`constraints-|version|`.

> This way, we keep a tested set of dependencies at the moment of release. This provides you with the ability
> of having the exact same installation of airflow + providers + dependencies as was known to be working
> at the moment of release - "golden" set of dependencies for that version of Airflow. There is a separate

Reviewer 2: I'd drop the golden piece. It makes it sound as overly desirable and useful in all cases, which may lead to some of the confusion we're trying to avoid.
```suggestion
This way, we keep a tested set of dependencies at the moment of release. This provides you with the ability
of having the exact same installation of airflow + providers + dependencies as was known to be working
at the moment of release. There is a separate
```

Author: Good idea. I think this could be why people are "clinging" to it. I changed it to "frozen" to get it a bit "chilled".

- ``PYTHON_VERSION`` Python version e.g. ``3.8``, ``3.9``

**However, that should not prevent you from being able to upgrade or downgrade providers and dependencies to other versions**

You can, for example, install new versions of providers and dependencies after the release to use the latest
version and up-to-date with latest security fixes - even if you do not want upgrade airflow core version.
Or you can downgrade some dependencies or providers if you want to keep previous versions for compatibility
reasons. Installing such dependencies should be done without constraints as a separate pip command.

When you do such an upgrade, you should make sure to also add the ``apache-airflow`` package to the list of
packages to install and pin it to the version that you have, otherwise you might end up with a
different version of Airflow than you expect as ``pip`` can upgrade/downgrade it automatically when
performing dependency resolution.

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"

>     pip install "apache-airflow[celery]==|version|" apache-airflow-providers-google==10.1.1

Author: ```suggestion
    pip install "apache-airflow==|version|" apache-airflow-providers-google==10.1.1
```

Such an upgrade is not always possible - you might have conflicting dependencies. However, by not using constraints,
you give ``pip`` a chance to resolve the conflicts for you, while keeping within the limits of dependencies
that Apache Airflow and other providers require. The combination of those dependencies might not be as well
tested as the "golden" set of dependencies, but it should work in most cases. You can also always run
the ``pip check`` command to test if the set of your Python packages is consistent and not conflicting.

Using your own constraints
==========================

When you decide to install your own dependencies, or want to upgrade or downgrade providers, you might want
to continue being able to keep reproducible installation of Airflow and those dependencies via a single command. In order to do that, you can produce your own constraints file and use it
to install Airflow instead of the one provided by the community.

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"

>     pip install "apache-airflow[celery]==|version|" dbt-core==0.20.0

Author: ```suggestion
    pip install "apache-airflow==|version|" dbt-core==0.20.0
```

- ``PYTHON_VERSION`` Python version e.g. ``3.8``, ``3.9``

**However, that should not prevent you from being able to upgrade or downgrade providers and dependencies to other versions**

You can, for example, install new versions of providers and dependencies after the release to use the latest
version and up-to-date with latest security fixes - even if you do not want upgrade airflow core version.
Or you can downgrade some dependencies or providers if you want to keep previous versions for compatibility
reasons. Installing such dependencies should be done without constraints as a separate pip command.

When you do such an upgrade, you should make sure to also add the ``apache-airflow`` package to the list of
packages to install and pin it to the version that you have, otherwise you might end up with a
different version of Airflow than you expect as ``pip`` can upgrade/downgrade it automatically when
performing dependency resolution.

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"

>     pip install "apache-airflow[celery]==|version|" apache-airflow-providers-google==10.1.1

Reviewer 3: Hmm, is it necessary to specify extras again?

Author: Ah... NO :) Copy&paste -> great spot !

You can also downgrade or upgrade other dependencies this way - even if they are not compatible with
those dependencies that are stored in the original constraints file:

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"
    pip install "apache-airflow[celery]==|version|" dbt-core==0.20.0

Such an upgrade is not always possible - you might have conflicting dependencies. However, by not using constraints,
you give ``pip`` a chance to resolve the conflicts for you, while keeping within the limits of dependencies
that Apache Airflow and other providers require. The combination of those dependencies might not be as well
tested as the "golden" set of dependencies, but it should work in most cases. You can also always run
the ``pip check`` command to test if the set of your Python packages is consistent and not conflicting.

Using your own constraints
==========================

When you decide to install your own dependencies, or want to upgrade or downgrade providers, you might want

> to continue being able to keep reproducible installation of Airflow and those dependencies via single command
> line . In order to do that, you can produce your own constraints file and use it

Reviewer 3: ```suggestion
to continue being able to keep reproducible installation of Airflow and those dependencies via a single command. In order to do that, you can produce your own constraints file and use it
```

to install Airflow instead of the one provided by the community.

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"
    pip install "apache-airflow[celery]==|version|" dbt-core==0.20.0
    pip freeze > my-constraints.txt

Then you can use it to create reproducible installations of your environment in a single operation via
a local constraints file:

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "my-constraints.txt"

> The constraint file might also be hosted via a webserver of your choice and made available for remote use:

Reviewer 3: I'm not sure we need this example, we have plenty of examples here of using it over http.

Author: I will just make a general statement for it - no example is needed indeed.

- ``AIRFLOW_VERSION`` - Airflow version (e.g. :subst-code:`|version|`) or ``main``, ``2-0``, for latest development version
- ``PYTHON_VERSION`` Python version e.g. ``3.8``, ``3.9``

However, that should not prevent you from being able to install providers and dependencies that were upgraded
after the release to keep your installation up-to-dated, even if you do not upgrade airflow core version.

> Installing such dependencies should be done without constraints as a separate pip command.

Reviewer 2: What do you think about bolding this line to really make it stand out and hit the point home?

Author: Yes. I also rephrased it a little.

You should make sure to also add the ``apache-airflow`` package to the list of packages to install and pin
it to the version that you have, otherwise you might end up with a different version of Airflow than you
expect as ``pip`` can upgrade/downgrade it automatically when performing dependency resolution.

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"
    pip install "apache-airflow[celery]==|version|" apache-airflow-providers-google==10.1.1

You can also downgrade providers or other dependencies this way if you need to (for example if you do not
want to upgrade to the latest version of a provider).

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"
    pip install "apache-airflow[celery]==|version|" dbt-core==0.20.0

> Not always such upgrade is possible - you might have conflicting dependencies, but by not using constraints,

Reviewer 2: Such an upgrade is not always possible - you might have conflicting dependencies, but by not using constraints,
```

you give ``pip`` a chance to resolve the conflicts for you, while keeping within the limits of dependencies
that Apache Airflow and other providers require. The combination of those dependencies might not be as well
tested as the "golden" set of dependencies, but it should work in most cases. You can also always run
the ``pip check`` command to test if the set of your Python packages is consistent and not conflicting.

> Using your own constraints

Reviewer 2: Love this section, it should help lots of folks :+1:

==========================

When you decide to install your own dependencies, or want to upgrade or downgrade providers, you might want
to continue being able to run reproducible installation of Airflow and those dependencies. In order to do
that you can produce your own constraints file and use it to install Airflow instead of the one provided
by the community.

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"
    pip install "apache-airflow[celery]==|version|" dbt-core==0.20.0
    pip freeze > my-constraints.txt

> Then you can use it to run reproducible installation of your environment in a single operation via local file:

Reviewer 2: ```suggestion
Then you can use it to create reproducible installations of your environment in a single operation via a local constraints file:
```

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"
    pip install "apache-airflow[celery]==|version|" dbt-core==0.20.0
    pip freeze > my-constraints.txt

Then you can use it to run reproducible installation of your environment in a single operation via local file:

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "my-constraints.txt"

The constraint file might also be hosted via a webserver of your choice and made available for remote use:

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://my-company.org/my-constraints.txt"

  https://raw.githubusercontent.com/apache/airflow/constraints-latest/constraints-3.8.txt

> Fixing Constraint files at release time

Reviewer 2: I think this header and the first paragraph are worth keeping to set expectations about the constraints files.

Author: Yes. That was a bit too much, but I simplified it, removed middles section (which was a bit bullshit), and referred to "use your own constraints" in case the users want to keep updated between the released.

- ``AIRFLOW_VERSION`` - Airflow version (e.g. :subst-code:`|version|`) or ``main``, ``2-0``, for latest development version
- ``PYTHON_VERSION`` Python version e.g. ``3.8``, ``3.9``

The examples below assume that you want to use install airflow in a reproducible way with the ``celery`` extra,
but you can pick your own set of extras and providers to install.

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"

.. note::

    The reproducible installation guarantees that this initial installation steps will always work for you -
    providing that you use the right Python version and that you have appropriate Operating System dependencies
    installed for the providers to be installed. Some of the providers require additional OS dependencies to

>     be installed such us ``build-essential`` in order to compile libraries, or for example database client

Reviewer: ```suggestion
    be installed such as ``build-essential`` in order to compile libraries, or for example database client
```

    libraries in case you install a database provider, etc.. You need to figure out which system dependencies
    you need when your installation fails and install them before retrying the installation.

Upgrading and installing dependencies (including providers)
===========================================================

**The reproducible installation above should not prevent you from being able to upgrade or downgrade
providers and other dependencies to other versions**

You can, for example, install new versions of providers and dependencies after the release to use the latest
version and up-to-date with latest security fixes - even if you do not want upgrade airflow core version.
Or you can downgrade some dependencies or providers if you want to keep previous versions for compatibility
reasons. Installing such dependencies should be done without constraints as a separate pip command.

When you do such an upgrade, you should make sure to also add the ``apache-airflow`` package to the list of
packages to install and pin it to the version that you have, otherwise you might end up with a

> different version of Airflow than you expect as ``pip`` can upgrade/downgrade it automatically when

Reviewer: ```suggestion
different version of Airflow than you expect because ``pip`` can upgrade/downgrade it automatically when
```

performing dependency resolution.

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"
    pip install "apache-airflow==|version|" apache-airflow-providers-google==10.1.1

You can also downgrade or upgrade other dependencies this way - even if they are not compatible with
those dependencies that are stored in the original constraints file:

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"
    pip install "apache-airflow[celery]==|version|" dbt-core==0.20.0

.. warning::

    Not all dependencies can be installed this way - you might have dependencies conflicting with basic
    requirements of Airflow or other dependencies installed in your system. However, by skipping constraints
    when you install or upgrade dependencies, you give ``pip`` a chance to resolve the conflicts for you,
    while keeping dependencies within the limits that Apache Airflow, providers and other dependencies require.

>     The resulting combination of those dependencies might not be tested together before as well as the set
>     of dependencies that come with constraints, but it should work in most cases as we usually add

Reviewer: ```suggestion
    The resulting combination of those dependencies and the set of dependencies that come with the 
    constraints might not be tested before, but it should work in most cases as we usually add
```

    requirements, when Airflow depends on particular versions of some dependencies. In cases you cannot
    install some dependencies in the same environment as Airflow - you can attempt to use other approaches.
    See :ref:`best practices for handling conflicting/complex Python dependencies <best_practices/handling_conflicting_complex_python_dependencies>`

Verifying installed dependencies
================================

You can also always run the ``pip check`` command to test if the set of your Python packages is
consistent and not conflicting.

.. code-block:: bash

    > pip check
    No broken requirements found.

When you see such message and the exit code from ``pip check`` is 0, you can be sure, that there are no
conflicting dependencies in your environment.

Using your own constraints
==========================

When you decide to install your own dependencies, or want to upgrade or downgrade providers, you might want

> to continue being able to keep reproducible installation of Airflow and those dependencies via a single command. In order to do that, you can produce your own constraints file and use it
> to install Airflow instead of the one provided by the community.

Reviewer: ```suggestion
to continue being able to keep reproducible installation of Airflow and those dependencies via a single command.
In order to do that, you can produce your own constraints file and use it to install Airflow instead of the 
one provided by the community.
```

Just suggesting a new line char as I see this line is relatively longer than all other lines :)

.. code-block:: bash

    pip install "apache-airflow[celery]==|version|" --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-|version|/constraints-3.8.txt"
    pip install "apache-airflow==|version|" dbt-core==0.20.0

>     pip freeze > my-constraints.txt

Reviewer: I like this section a lot 👍🏽

Author: This is what a few people at the London meetup told as well. For me that was pretty obvious (It's how constraints were generated from the very beginning) but with the recent discussion and the talk I gave I realized that It was only "my bubble". People were - apparently so far- manually modifying the constraints and complained for example that recent google provider is not compatible with their (it turned out) custom constraints. 

Funny how some things obvious to you might be completely alien to others :)

    pip install "apache-airflow==|version|" "apache-airflow-providers-google==8.0.0"

.. note::

    Installing, upgrading, downgrading providers separately is not guaranteed to work with all
    Airflow versions or other providers. Some providers have minimum-required version of Airflow and some

>     versions of providers might have conflicting requirements with Airflow or other dependencies you

Reviewer: could we please improvise this statement "Some versions of providers might have conflicting requirements with Airflow" a bit? 

Do we mean some versions of Airflow here or we meant conflicting requirements with other Airflow providers?
Otherwise, this might be questioned if it is an Airflow provider why or in which cases it would have conflicting requirements with Airflow?

Author: Good point. I added a very specific recent example that illustrates it (google provider + google ads).

Reviewer: yes, pretty clear and detailed now. Thank you!
